src/utils/OnboardingProvider.tsx

```diff
@@ -0,0 +1,293 @@
+'use client';
+
+import supabase from '@/api/supabase/createClient';
+import {
+  deleteLanguages,
+  deleteRoles,
+  upsertLanguages,
+  upsertProfile,
+  upsertRoles,
+} from '@/api/supabase/queries/profiles';
+import { Profile, ProfileLanguage, ProfileRole } from '@/types/schema';
+import { UUID } from 'crypto';
+import {
+  createContext,
+  useState,
+  useMemo,
+  ReactNode,
+  SetStateAction,
+  Dispatch,
+  useEffect,
+} from 'react';
+
+interface FlowData {
+  url: string;
+  name: string;
+}
+
+interface OnboardingContextType {
+  profile: Profile;
+  languages: ProfileLanguage[];
+  roles: ProfileRole[];
+  progress: number;
+  flow: FlowData[];
+  canContinue: boolean;
+  flushProfileData: () => Promise<void>;
+  updateProfile: (updateInfo: Partial<Profile>) => Promise<void>;
+  setProgress: Dispatch<SetStateAction<number>>;
+  setCanContinue: Dispatch<SetStateAction<boolean>>;
+  setFlow: Dispatch<SetStateAction<FlowData[]>>;
+  setLanguages: (languages: ProfileLanguage[]) => Promise<void>;
+  setRoles: (roles: ProfileRole[]) => Promise<void>;
+}
+
+export const OnboardingContext = createContext<
+  OnboardingContextType | undefined
+>(undefined);
+
+const blankProfile: Profile = {
+  first_name: '',
+  last_name: '',
+  hours_per_month: 0,
+  location: '',
+  start_date: '',
+  user_id: '0-0-0-0-0',
+};
+
+export default function OnboardingProvider({
+  children,
+}: {
+  children: ReactNode;
+}) {
+  const [progress, setProgress] = useState(0);
+  const [flow, setFlow] = useState<FlowData[]>([]);
+  const [profile, setProfile] = useState<Profile>({ ...blankProfile });
+  const [languages, setProfileLangs] = useState<ProfileLanguage[]>([]);
+  const [roles, setProfileRoles] = useState<ProfileRole[]>([]);
+  const [canContinue, setCanContinue] = useState<boolean>(false);
+
+  useEffect(() => {
+    (async () => {
+      const { data: sessionData, error } = await supabase.auth.getSession();
+
+      if (error) throw error;
+
+      const userId = sessionData?.session?.user.id;
+      if (userId) {
+        setProfile({ ...blankProfile, user_id: userId as UUID });
+      }
+    })();
+  }, []);
+
+  const providerValue = useMemo(() => {
+    /**
+     * Updates stored profile state with the partial data.
+     * Does not affect database.
+     */
+    const updateProfile = async (updatedInfo: Partial<Profile>) => {
+      const newProfileData = { ...profile, ...updatedInfo };
+      setProfile(newProfileData);
+    };
+
```

Comment:
can we create another function here (perhaps called `completeProfile` or `finishProfile` or something along those lines) that updates the Profile here and then also updates the profile in the Profile context (which would then update the row in the `profiles` table on supabase?)

i'm thinking we would use this for the last page of the onboarding flow

Reply:
Can't we call both..?

Reply:
oh lol you're right, we can just access the data from the onboarding context and pass this into the profile context from the final page....ok sounds good!